test/AuthEnabledVDMSServer.cc

```diff
@@ -0,0 +1,179 @@
+
+std::string random_string(size_t length)
+{
+    auto randchar = []() -> char
+    {
+        const char charset[] =
+            "0123456789"
+            "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+            "abcdefghijklmnopqrstuvwxyz";
+
+        const size_t max_index = sizeof(charset) - 1;
+
+        return charset[rand() % max_index];
+    };
+
+    std::string str(length, 0);
+    std::generate_n(str.begin(), length, randchar);
+
+    return str;
+}
+
+AuthEnabledVDMSServer::AuthEnabledVDMSServer(int port, AuthEnabledVDMSServerConfig config) :
+    _server(new comm::ConnServer(port, config.connServerConfig))
+{
+    auto thread_function = [&]()
+    {
+        std::shared_ptr<comm::Connection> server_conn = _server->accept();
+
+        while (!_stop_signal) {
+            protobufs::queryMessage protobuf_request;
+
+            try {
+                protobuf_request = receive_message(server_conn);
+            } catch (comm::Exception exception) {
+                if (exception.num == comm::ConnectionShutDown) {
+                    return;
+                }
+
+                throw exception;
+            }
+
+            protobufs::queryMessage protobuf_response;
+
+            auto is_authenticate_request_ = is_authenticate_request(protobuf_request);
+            auto is_refresh_token_request_ = is_refresh_token_request(protobuf_request);
+
+            if (is_authenticate_request_ || is_refresh_token_request_) {
+                regenerate_tokens();
+
+                auto command_name = is_authenticate_request_ ? "Authenticate" : "RefreshToken";
+
+                auto responseJson = nlohmann::json::array({{
+                    {command_name, {
+                        {"session_token", session_token},
+                        {"session_token_expires_in", config.session_token_expires_in},
+                        {"refresh_token", refresh_token},
+                        {"refresh_token_expires_in", config.refresh_token_expires_in},
+                        {"status", 0}
+                    }}
+                }});
+
+                protobuf_response.set_json(responseJson.dump());
+            }
+            else {
+                if (protobuf_request.token() == session_token) {
+                    protobuf_response.set_json(protobuf_request.json());
+                }
+                else {
+                    auto responseJson = nlohmann::json::array({{
+                        {"Something", {
+                            {"status", -1}
+                        }}
+                    }});
+
+                    protobuf_response.set_json(responseJson.dump());
+                }
+            }
+
+            send_message(server_conn, protobuf_response);
+        }
+    };
+
+    _work_thread = std::unique_ptr<std::thread>(new std::thread(thread_function));
+}
+
+AuthEnabledVDMSServer::~AuthEnabledVDMSServer()
+{
+    _stop_signal = true;
+
+    if (_work_thread->joinable()) {
+        _work_thread->join();
+    }
```

Comment:
because this is testing code, and we use gtest, this is not that important because the execution of the tests will continue and exception are catch by the testing framework.

Reply:
> because this is testing code, and we use gtest, this is not that important because the execution of the tests will continue and exception are catch by the testing framework.

This is a destructor, so I don't think it is safe to assume this.  If this dtor throws during stack unwinding from another exception, the process will immediately terminate.  I think also in C++11 dtors are implicitly `noexcept`, and the process will terminate when a dtor throws even without the context of another exception (at least on some platforms).

My understanding is that the general best practice is never to throw from a destructor unless you have a very good reason.  Relevant article: https://akrzemi1.wordpress.com/2011/09/21/destructors-that-throw/